main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1081 @@
+        // resolved BEFORE purifying the region as we exit
+        c.enterRegionM(regionVar.sym)
+        c.unifyTypeM(sym.tvar, Type.mkRegion(regionVar, loc), loc)
+        val (tpe, eff) = visitExp(exp)
+        c.exitRegionM(evar, eff, loc)
+        val resTpe = tpe
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Match(exp, rules, loc) =>
+        val patterns = rules.map(_.pat)
+        val guards = rules.flatMap(_.guard)
+        val bodies = rules.map(_.exp)
+        val guardLocs = guards.map(_.loc)
+        val (tpe, eff) = visitExp(exp)
+        val patternTypes = patterns.map(visitPattern)
+        c.unifyAllTypesM(tpe :: patternTypes, Kind.Star, loc)
+        val (guardTpes, guardEffs) = guards.map(visitExp).unzip
+        guardTpes.zip(guardLocs).foreach { case (gTpe, gLoc) => c.expectTypeM(expected = Type.Bool, actual = gTpe, loc = gLoc) }
+        val (bodyTypes, bodyEffs) = bodies.map(visitExp).unzip
+        val resTpe = c.unifyAllTypesM(bodyTypes, Kind.Star, loc)
+        val resEff = Type.mkUnion(eff :: guardEffs ::: bodyEffs, loc)
+        (resTpe, resEff)
+
+      case Expr.TypeMatch(exp, rules, loc) =>
+        val bodies = rules.map(_.exp)
+        val (_, eff) = visitExp(exp)
+        // rigidify all the type vars in the rules
+        rules.flatMap(_.tpe.typeVars.toList).map(_.sym).foreach(c.rigidifyM)
+        // unify each rule's variable with its type
+        rules.foreach { rule => c.unifyTypeM(rule.sym.tvar, rule.tpe, rule.sym.loc) }
+        val (bodyTypes, bodyEffs) = bodies.map(visitExp).unzip
+        val resTpe = c.unifyAllTypesM(bodyTypes, Kind.Star, loc)
+        val resEff = Type.mkUnion(eff :: bodyEffs, loc)
+        (resTpe, resEff)
+
+      case e: Expr.RestrictableChoose => RestrictableChooseConstraintGeneration.visitRestrictableChoose(e)
+
+      case KindedAst.Expr.Tag(symUse, exp, tvar, loc) =>
+        // Lookup the enum declaration.
+        val decl = root.enums(symUse.sym.enumSym)
+
+        // Lookup the case declaration.
+        val caze = decl.cases(symUse.sym)
+
+        // Instantiate the type scheme of the case.
+        val (_, tagType) = Scheme.instantiate(caze.sc, loc.asSynthetic)
+
+        //
+        // The tag type is a function from the type of variant to the type of the enum.
+        //
+        val (tpe, eff) = visitExp(exp)
+        c.unifyTypeM(tagType, Type.mkPureArrow(tpe, tvar, loc), loc)
+        val resTpe = tvar
+        val resEff = eff
+        (resTpe, resEff)
+
+      case e: Expr.RestrictableTag => RestrictableChooseConstraintGeneration.visitRestrictableTag(e)
+
+      case Expr.Tuple(elms, loc) =>
+        val (elmTpes, elmEffs) = elms.map(visitExp).unzip
+        val resTpe = Type.mkTuple(elmTpes, loc)
+        val resEff = Type.mkUnion(elmEffs, loc)
+        (resTpe, resEff)
+
+      case Expr.RecordEmpty(loc) =>
+        val resTpe = Type.mkRecord(Type.RecordRowEmpty, loc)
```

Comment:
`Type.RecordRowEmpty` should be `Type.mkRecordRowEmpty(loc)` right? same thing in the next case

Reply:
eh, it creates an extra object for no real gain (type locations for inferred types are evil)

Reply:
I agree with Matt